presto-main/src/main/java/com/facebook/presto/operator/scalar/DayOfMonthWithEnglishSuffixPrinter.java

```diff
@@ -0,0 +1,192 @@
+    @Override
+    public int estimatePrintedLength()
+    {
+        return MAX_PRINTED_LENGTH;
+    }
+
+    private static String getSuffix(int day)
+    {
+        Preconditions.checkArgument(day > 0 && day <= 31, "%s is not a valid day of month", day);
+        switch(day) {
+            case 1:
+            case 21:
+            case 31:
+                return "st";
+            case 2:
+            case 22:
+                return "nd";
+            case 3:
+            case 23:
+                return "rd";
+            default:
+                return "th";
+        }
+    }
+
+    private void printTo(Appendable appendable, long instant, Chronology chrono) throws IOException
+    {
+        int day = DAY_OF_MONTH.getField(chrono).get(instant);
+        FormatUtils.appendUnpaddedInteger(appendable, day);
+        appendable.append(getSuffix(day));
+    }
+
+    @Override
+    public void printTo(StringBuffer buf, long instant, Chronology chrono, int displayOffset, DateTimeZone displayZone, Locale locale)
+    {
+        try {
+            printTo(buf, instant, chrono);
+        }
+        catch (IOException e) {
+            // StringBuffer does not throw IOException
+        }
+    }
+
+    @Override
+    public void printTo(Writer out, long instant, Chronology chrono, int displayOffset, DateTimeZone displayZone, Locale locale)
+            throws IOException
+    {
+        printTo(out, instant, chrono);
+    }
+
+    private void printTo(Appendable appendable, ReadablePartial partial) throws IOException
+    {
+        if (partial.isSupported(DAY_OF_MONTH)) {
+            int day = partial.get(DAY_OF_MONTH);
+            FormatUtils.appendUnpaddedInteger(appendable, day);
+            appendable.append(getSuffix(day));
+        }
+        else {
+            appendUnknownString(appendable, MIN_PRINTED_LENGTH);
+        }
+    }
+
+    @Override
+    public void printTo(StringBuffer buf, ReadablePartial partial, Locale locale)
+    {
+        try {
+            printTo(buf, partial);
+        }
+        catch (IOException ex) {
+            // StringBuffer does not throw IOException
+        }
+    }
+
+    @Override
+    public void printTo(Writer out, ReadablePartial partial, Locale locale)
+            throws IOException
+    {
+        printTo(out, partial);
+    }
+
+    private static void appendUnknownString(Appendable appendable, int len) throws IOException
+    {
+        for (int i = len; i >= 0; i--) {
+            appendable.append('\ufffd');
+        }
+    }
+
+    @Override
+    public int estimateParsedLength()
+    {
+        return MAX_PRINTED_LENGTH;
+    }
+
+    @Override
+    public int parseInto(DateTimeParserBucket bucket, String text, int position)
+    {
+        if (text.length() - position < MIN_PRINTED_LENGTH) {
+            return ~position;
+        }
+
+        String token = text.substring(position, position + MIN_PRINTED_LENGTH).toLowerCase();
+        if (THREE_LETTER_PARSE_LUT.containsKey(token)) {
```

Comment:
It may be easier to follow in this specific case because the declaration is above in this class. In general, it may require understanding non-local context to be able to reason about the code. ImmutableMap is the chosen implementation, but the type of the field is Map and this code interacts with the Map interface. It's not sufficient to look at this code to know how it's expected to behave (leaky abstractions, etc.). Now, if the type of the field/variable were ImmutableMap, it'd be another story, since it's documented as not supporting null values.